File: Leetcode/python-solutions/max_profit.py
class Solution(object):
    def maxProfit(self, prices):
        """
        :type prices: List[int]
        :rtype: int
        """

        n = len(prices)

        if n < 2:
            return 0

        min_price = prices[0]
        max_profit = 0

        for price in prices[1:]:
            if price < min_price:
                min_price = price
            else:
                max_profit = max(max_profit, price - min_price)
        return max_profit

File: Leetcode/typescript/max-profit.ts
function maxProfit(prices: number[]): number {
  let l = 0;
  let r = 1;
  let profit = 0;

  while (r < prices.length) {
    if (prices[r] < prices[l]) {
      [l, r] = [r, r + 1];
    } else {
      profit = Math.max(profit, prices[r] - prices[l]);
      r += 1;
    }
  }
  return profit;
}
